src/tests/utils/test_cache_logger.py

```diff
@@ -0,0 +1,226 @@
+        self.assertIn("GetSnoopResult: Address 5678", stderr_output)
+        self.assertIn("Snoop Result: HIT", stderr_output)
+
+        # Clear buffer
+        self.stderr.seek(0)
+        self.stderr.truncate()
+
+        # Test put snoop result
+        dummy.PutSnoopResult(0x9ABC, SnoopResult.NOHIT)
+        stderr_output = self.stderr.getvalue()
+        self.assertIn("SnoopResult: Address 9abc", stderr_output)
+        self.assertIn("SnoopResult: NOHIT", stderr_output)
+
+    def test_message_to_cache_logging(self):
+        """Test logging of cache messages"""
+        dummy = self.create_dummy_cache_with_logger(self.logger)
+
+        dummy.MessageToCache(CacheMessage.INVALIDATELINE, 0xDEF0)
+        stderr_output = self.stderr.getvalue()
+        self.assertIn("L2: INVALIDATELINE def0", stderr_output)
+
+    def test_print_stats(self):
+        """Test statistics printing"""
+        self.logger.cache_reads = 100
+        self.logger.cache_writes = 50
+        self.logger.cache_hits = 75
+        self.logger.cache_misses = 25
+
+        self.logger.print_stats(self.stdout)
+        stats_output = self.stdout.getvalue()
+
+        self.assertIn("Number of cache reads:  100", stats_output)
+        self.assertIn("Number of cache writes: 50", stats_output)
+        self.assertIn("Number of cache hits:   75", stats_output)
+        self.assertIn("Number of cache misses: 25", stats_output)
+        self.assertIn("Cache hit ratio:        75.00000", stats_output)
+
+    def test_logger_output_demonstration(self):
+        """Demonstration test that shows actual logger output"""
+        # Create a logger that writes to both StringIO (for assertions) and console
+        import sys
+
+        demo_logger = CacheLogger(
+            level=LogLevel.DEBUG,
+            stdout=sys.stdout,  # Write to console
+            stderr=sys.stderr,  # Write to console
+        )
+
+        print("\n=== Logger Output Demonstration ===")
+        print("This test shows actual logger output at different levels:\n")
+
+        # Create dummy cache with this logger
+        dummy = self.create_dummy_cache_with_logger(demo_logger)
+
+        print("1. Testing Bus Operation logging:")
+        dummy.BusOperation(BusOp.READ, 0x1234, SnoopResult.NOHIT)
+
+        print("\n2. Testing Snoop Result logging:")
+        dummy.GetSnoopResult(0x5678)
+        dummy.PutSnoopResult(0x9ABC, SnoopResult.NOHIT)
+
+        print("\n3. Testing Cache Message logging:")
+        dummy.MessageToCache(CacheMessage.INVALIDATELINE, 0xDEF0)
+
+        print("\n4. Testing Statistics:")
+        demo_logger.cache_reads = 100
+        demo_logger.cache_writes = 50
+        demo_logger.cache_hits = 75
+        demo_logger.cache_misses = 25
+        demo_logger.print_stats()
+
+        print("\n=== End Logger Output Demonstration ===")
+
+    def test_file_logging(self):
+        """Demonstration of logging to files. Creates debug.log and stats.log in the
+        tests/utils directory."""
```

Comment:
Bonus points for an integration test!

I might vote we have the test clean up after itself and remove the files, keeping them out of the repo, since the test would fail if it fails to create the files right? That way if we change the test, we don't have diffs on test output. But I don't think it's a big problem if they are checked in.

Reply:
I agree with this, its generally good practice for the test to clean up after but I left them for when we write our report.